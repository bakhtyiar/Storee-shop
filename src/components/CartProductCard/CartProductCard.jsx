import React, {useContext} from 'react';
import {RootContext} from "../../contexts/root-context";
import {Link} from "react-router-dom";
import {routes} from "../../utils/constants";
import imageIcon from "../../assets/img/icons/image/icons8-image-48.png";
import {Button} from "react-bootstrap";
import styled from "styled-components";

const StyledImg = styled.img`
  align-self: center;
  max-width: 100%;
  max-height: 200px;
`;

const CartProductCard = ({product, selfIndexInCart}) => {
    const { cartState: { onRemoveFromCart } } = useContext(RootContext);

    return (
        <Link className='px-2 py-3 border rounded d-flex flex-column align-items-start text-decoration-none text-black' to={`${routes.product.path}/${product.id}`}>
            <StyledImg src={imageIcon} alt="image icon"/>
            <h6 className='mt-2'>{product.title}</h6>
            <p>Price: {product.price}</p>
            <p>Amount: {product.quantity}</p>
            <Button
                variant='dark'
                className='w-100'
                onClick={ (e) => {
                    e.preventDefault();
                    onRemoveFromCart(selfIndexInCart);
                }}
            >
                <i className="bi bi-x"></i>
                {' '}
                Remove from cart
            </Button>
        </Link>
    );
};

export default CartProductCard;