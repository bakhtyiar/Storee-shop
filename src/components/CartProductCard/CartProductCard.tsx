import React, {useContext, useEffect, useState} from 'react';
import {RootContext} from "../../contexts/root-context/root-context";
import {Link} from "react-router-dom";
import {routes} from "../../utils/constants";
import imageIcon from "../../assets/img/icons/image/icons8-image-48.png";
import {Button, Form} from "react-bootstrap";

const CartProductCard = ({
    product,
    selfIndexInCart
}: any) => {
    const {cartState: {onRemoveFromCart, onUpdateQuantity}} = useContext(RootContext);
    const [quantity, setQuantity] = useState(product.quantity);

    useEffect(() => {
        if (quantity < 1) {
            onRemoveFromCart(selfIndexInCart);
        } else {
            onUpdateQuantity(product.id, quantity);
        }
    }, [quantity]);

    return (
        <div className='px-2 py-3 border rounded d-flex flex-row align-items-center text-decoration-none text-black' data-testid='cart-product-card'>
            
            <Link className='w-100 d-flex flex-row text-decoration-none text-black'
                  to={`${routes.product.path}/${product.id}`}
                  data-testid='product-link'
            >
                
                <div className={'bg-image rounded align-self-center'}
                     style={{
                         height: '64px',
                         width: '64px',
                         backgroundImage: `url(${product.thumbnail || imageIcon})`,
                         backgroundPosition: 'center',
                         backgroundSize: 'contain',
                         backgroundRepeat: 'no-repeat'
                     }}/>
                
                <div className='flex-column ms-2'>
                    
                    <h6 className='mt-2'>{product.title}</h6>
                    
                    <p className='my-0'>Price: {product.price}</p>
                </div>
            </Link>
            
            <div className='ml-3 d-flex gap-2 flex-column flex-lg-row'>
                
                <div className='d-flex'>
                    
                    <Button
                        variant="outline-dark"
                        onClick={() => {
                            setQuantity((prevState: any) => prevState - 1);
                        }}
                        data-testid="minus-btn"
                    >
                        -
                    </Button>
                    
                    <Form.Control
                        type="number"
                        value={quantity}
                        placeholder="Quantity"
                        onChange={(e) => {
                            setQuantity(Number(e.target.value));
                        }}
                        data-testid={"quantity-field"}
                    />
                    
                    <Button
                        variant="outline-dark"
                        onClick={() => {
                            setQuantity((prevState: any) => prevState + 1);
                        }}
                        data-testid="plus-btn"
                    >
                        +
                    </Button>
                </div>
                
                <Button
                    variant='outline-dark'
                    className='d-flex rounded justify-content-center'
                    onClick={(e) => {
                        e.preventDefault();
                        onRemoveFromCart(selfIndexInCart);
                    }}
                    data-testid="remove-btn"
                >
                    
                    <i className="bi bi-x"></i>
                    {' '}
                    Remove
                </Button>
            </div>
        </div>
    );
};

export default CartProductCard;