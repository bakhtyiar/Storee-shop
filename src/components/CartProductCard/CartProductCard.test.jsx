import CartProductCard from "./CartProductCard";
import {screen} from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import {renderWithRouter} from "../../tests/helpers/renderWithRouter";

describe("CartProductCard tests", () => {
    test("Match snapshot", () => {
        renderWithRouter(<CartProductCard product={{quantity: 1}} selfIndexInCart={0}/>);
        expect(screen.getByTestId("cart-product-card")).toMatchSnapshot();
    })
    test("Minus quantity", () => {
        renderWithRouter(<CartProductCard product={{quantity: 2}} selfIndexInCart={0}/>);
        let cartProductCard = screen.getByTestId("cart-product-card");
        expect(cartProductCard).toBeInTheDocument();
        let minusButton = screen.getByTestId("minus-btn");
        expect(minusButton).toBeInTheDocument();
        userEvent.click(minusButton);
        let quantityField = screen.getByTestId("quantity-field");
        expect(quantityField.value).toBe("1");
    })
    test("Change quantity", () => {
        renderWithRouter(<CartProductCard product={{quantity: 1}} selfIndexInCart={0}/>);
        let cartProductCard = screen.getByTestId("cart-product-card");
        expect(cartProductCard).toBeInTheDocument();
        let quantityField = screen.getByTestId("quantity-field");
        expect(quantityField).toBeInTheDocument();
    })
    test("Plus quantity", () => {
        renderWithRouter(<CartProductCard product={{quantity: 1}} selfIndexInCart={0}/>);
        let cartProductCard = screen.getByTestId("cart-product-card");
        expect(cartProductCard).toBeInTheDocument();
        let plusButton = screen.getByTestId("plus-btn");
        expect(plusButton).toBeInTheDocument();
        userEvent.click(plusButton);
        let quantityField = screen.getByTestId("quantity-field");
        expect(quantityField).toBeInTheDocument();
        expect(quantityField.value).toBe("2");
    })
    test("Remove", () => {
        renderWithRouter(<CartProductCard product={{quantity: 1}} selfIndexInCart={0}/>);
        let removeButton = screen.getByTestId("remove-btn");
        expect(removeButton).toBeInTheDocument();
    })
})