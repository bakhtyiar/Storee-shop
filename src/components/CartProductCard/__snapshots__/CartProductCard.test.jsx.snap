// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartProductCard test Match snapshot 1`] = `
Object {
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "logTestingPlaygroundURL": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
}
`;

exports[`CartProductCard tests Match snapshot 1`] = `
<div
  class="px-2 py-3 border rounded d-flex flex-row align-items-center text-decoration-none text-black"
  data-testid="cart-product-card"
>
  <a
    class="w-100 d-flex flex-row text-decoration-none text-black"
    href="/product/undefined"
  >
    <div
      class="bg-image rounded align-self-center"
      style="height: 64px; width: 64px; background-image: url(icons8-image-48.png); background-position: center; background-size: contain; background-repeat: no-repeat;"
    />
    <div
      class="flex-column ms-2"
    >
      <h6
        class="mt-2"
      />
      <p
        class="my-0"
      >
        Price: 
      </p>
    </div>
  </a>
  <div
    class="ml-3 d-flex gap-2"
  >
    <div
      class="d-flex"
    >
      <button
        class="btn btn-outline-dark"
        data-testid="minus-btn"
        type="button"
      >
        -
      </button>
      <input
        class="form-control"
        data-testid="quantity-field"
        placeholder="Quantity"
        type="number"
        value="1"
      />
      <button
        class="btn btn-outline-dark"
        data-testid="plus-btn"
        type="button"
      >
        +
      </button>
    </div>
    <button
      class="d-flex rounded btn btn-outline-dark"
      data-testid="remove-btn"
      type="button"
    >
      <i
        class="bi bi-x"
      />
       
      Remove
    </button>
  </div>
</div>
`;
